Fusion threads (#9405)
fishtown-analytics/docs.getdbt.com: website/docs/docs/running-a-dbt-project/using-threads.md

You can use a different number of threads than the value defined in your target by using the `--threads` option when executing a dbt command.

You will define the number of threads in your `profiles.yml` file (when developing locally with dbt Core and the dbt Fusion engine), <Constant name="cloud" /> job definition, and <Constant name="cloud" /> development credentials under your profile.

<VersionBlock firstVersion="1.12">
<!-- versioning for 1.12 so it shows up in Latest eventually we might want to firstVersion="2.0"-->

## Fusion engine thread optimization

> In the context of <Constant name="fusion"/>, a **thread** is an open connection to your data warehouse—not the number of parallel threads on your local machine's CPU. Data platforms vary in how many concurrent connections they allow; exceeding those limits causes the platform to reject new connections.

Author: ```suggestion
In the context of <Constant name="fusion"/>, a **thread** is an open connection to your data warehouse, not the number of parallel threads on your local machine's CPU. Data platforms vary in how many concurrent connections they allow; exceeding those limits causes the platform to reject new connections.
```

> In the context of <Constant name="fusion"/>, a **thread** is an open connection to your data warehouse, not the number of parallel threads on your local machine's CPU. Data platforms vary in how many concurrent connections they allow; exceeding those limits causes the platform to reject new connections.

Reviewer: Hiya Matt

As thread isn't a UI element, maybe this should be backticked or italicised? I checked the other docs and it looks like it should be backticked (there are some instances of plain text) but maybe italicised for emphasis?

Reviewer: Also (maybe just me), this reads quite technical to me. Maybe it could be simplified to something like:

"In the context of <Constant name="fusion"/>, `threads` caps the maximum number of concurrent connections to your data warehouse, not parallel threads on your local machine’s CPU. Connections are pooled and reused, so `threads` is a limit on concurrency, not a fixed count of open connections. Data platforms vary in how many concurrent connections they allow; exceeding those limits can cause connection errors, queuing, or timeouts."

> In the context of <Constant name="fusion"/>, a **thread** is an open connection to your data warehouse, not the number of parallel threads on your local machine's CPU. Data platforms vary in how many concurrent connections they allow; exceeding those limits causes the platform to reject new connections.

Author: ```suggestion
In the context of <Constant name="fusion"/>, a thread is an open connection to your data warehouse, not the number of parallel threads on your local machine's CPU. Data platforms vary in how many concurrent connections they allow; exceeding those limits causes the platform to reject new connections.
```

> Historically, analytics engineers set `threads:` to ensure dbt never opened more connections than the platform could handle.

Reviewer: Hiya Matt

Small nit: `threads:` reads a bit awkward with the colon inside backticks, maybe just `threads` or `threads: 4` in profiles.yml?

Author: Understood but in this context, `threads:` is the syntax that would be used to set the configutration. Including the 4 or any other number might give the wrong impression in this case

> <Constant name="fusion"/> works well without configuring `threads`. Rather than treating `threads` as a strict limit, <Constant name="fusion"/> automatically manages connection parallelism based on platform limits and uses backpressure to avoid overloading your warehouse. In general, we recommend not setting `threads` when using <Constant name="fusion"/>.

Reviewer: Small suggestion, "backpressure" might be a bit too jargon-y for some users, maybe it could be simplified?

Author: I totally agree here, but it is the term widely used to describe what's happening, so it should make sense to people who need threads

However, if <Constant name="fusion"/> still opens more connections than your warehouse can handle, configure `threads` to cap the maximum number of concurrent connections.

> To run one operation at a time, use the `--no-parallel` flag. This overrides the configured thread count entirely and is useful for debugging.

Reviewer: Hiya Matt

I might be misunderstanding this, but is the second sentence still accurate?

I had a look at the [dbt-fusion CHANGELOG.md](https://github.com/dbt-labs/dbt-fusion/blob/main/CHANGELOG.md) it seems that `--no-parallel` and threads have been decoupled. One of the changelog entries mentions moving tests away from `--threads=1` for sequential runs because it was limiting the connection pool rather than controlling execution order.

If that's the case, would `--threads 1` (or a low thread count in profiles.yml) be a better recommendation for troubleshooting connection issues, since threads still appears to control connection limits?

Also, I noticed `--no-parallel` is currently hidden in the CLI help, so it might be worth confirming whether we want to document it here.

Happy to be corrected if I'm missing some context!

Author: Yeah this can be worded better. WIll fix

> To run one operation at a time, use the `--no-parallel` flag. This overrides the configured thread count entirely and is useful for debugging.

Author: ```suggestion
Project parsing runs separately and automatically uses all available CPUs. To disable parallel parsing and run one operation at a time, use the `--no-parallel flag`. This is useful for debugging parse errors and does not affect threads.
```

> Project parsing runs separately and automatically uses all available CPUs. To disable parallel parsing and run one operation at a time, use the `--no-parallel flag`. This is useful for debugging parse errors and does not affect threads.

Reviewer: ```suggestion
Project parsing runs separately and automatically uses all available CPUs. To disable parallel parsing and run one operation at a time, use the `--no-parallel` flag. This is useful for debugging parse errors and does not affect threads.
```
